File: airbyte-integrations/connectors/destination-gcs/src/main/java/io/airbyte/integrations/destination/gcs/jsonl/GcsJsonlWriter.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.gcs.jsonl;

import alex.mojaki.s3upload.MultiPartOutputStream;
import alex.mojaki.s3upload.StreamTransferManager;
import com.amazonaws.services.s3.AmazonS3;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.airbyte.commons.jackson.MoreMappers;
import io.airbyte.commons.json.Jsons;
import io.airbyte.integrations.base.JavaBaseConstants;
import io.airbyte.integrations.destination.gcs.GcsDestinationConfig;
import io.airbyte.integrations.destination.gcs.writer.BaseGcsWriter;
import io.airbyte.integrations.destination.s3.S3Format;
import io.airbyte.integrations.destination.s3.util.StreamTransferManagerFactory;
import io.airbyte.integrations.destination.s3.writer.DestinationFileWriter;
import io.airbyte.protocol.models.v0.AirbyteRecordMessage;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.sql.Timestamp;
import java.util.UUID;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class GcsJsonlWriter extends BaseGcsWriter implements DestinationFileWriter {

  protected static final Logger LOGGER = LoggerFactory.getLogger(GcsJsonlWriter.class);

  private static final ObjectMapper MAPPER = MoreMappers.initMapper();

  private final StreamTransferManager uploadManager;
  private final MultiPartOutputStream outputStream;
  private final PrintWriter printWriter;
  private final String gcsFileLocation;
  private final String objectKey;

  public GcsJsonlWriter(final GcsDestinationConfig config,
                        final AmazonS3 s3Client,
                        final ConfiguredAirbyteStream configuredStream,
                        final Timestamp uploadTimestamp) {
    super(config, s3Client, configuredStream);

    final String outputFilename = BaseGcsWriter.getOutputFilename(uploadTimestamp, S3Format.JSONL);
    objectKey = String.join("/", outputPrefix, outputFilename);

    gcsFileLocation = String.format("gs://%s/%s", config.getBucketName(), objectKey);
    LOGGER.info("Full GCS path for stream '{}': {}/{}", stream.getName(), config.getBucketName(), objectKey);

    this.uploadManager = StreamTransferManagerFactory
        .create(config.getBucketName(), objectKey, s3Client)
        .get();

    // We only need one output stream as we only have one input stream. This is reasonably performant.
    this.outputStream = uploadManager.getMultiPartOutputStreams().get(0);
    this.printWriter = new PrintWriter(outputStream, true, StandardCharsets.UTF_8);
  }

  @Override
  public void write(final UUID id, final AirbyteRecordMessage recordMessage) {
    final ObjectNode json = MAPPER.createObjectNode();
    json.put(JavaBaseConstants.COLUMN_NAME_AB_ID, id.toString());
    json.put(JavaBaseConstants.COLUMN_NAME_EMITTED_AT, recordMessage.getEmittedAt());
    json.set(JavaBaseConstants.COLUMN_NAME_DATA, recordMessage.getData());
    printWriter.println(Jsons.serialize(json));
  }

  @Override
  public void write(final JsonNode formattedData) throws IOException {
    printWriter.println(Jsons.serialize(formattedData));
  }

  @Override
  protected void closeWhenSucceed() {
    printWriter.close();
    outputStream.close();
    uploadManager.complete();
  }

  @Override
  protected void closeWhenFail() {
    printWriter.close();
    outputStream.close();
    uploadManager.abort();
  }

  @Override
  public String getFileLocation() {
    return gcsFileLocation;
  }

  @Override
  public S3Format getFileFormat() {
    return S3Format.JSONL;
  }

  @Override
  public String getOutputPath() {
    return objectKey;
  }

}

File: airbyte-cdk/java/airbyte-cdk/gcs-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/gcs/jsonl/GcsJsonlWriter.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.destination.gcs.jsonl

import alex.mojaki.s3upload.MultiPartOutputStream
import alex.mojaki.s3upload.StreamTransferManager
import com.amazonaws.services.s3.AmazonS3
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import io.airbyte.cdk.integrations.base.JavaBaseConstants
import io.airbyte.cdk.integrations.destination.gcs.GcsDestinationConfig
import io.airbyte.cdk.integrations.destination.gcs.writer.BaseGcsWriter
import io.airbyte.cdk.integrations.destination.s3.FileUploadFormat
import io.airbyte.cdk.integrations.destination.s3.util.StreamTransferManagerFactory.create
import io.airbyte.cdk.integrations.destination.s3.writer.DestinationFileWriter
import io.airbyte.commons.jackson.MoreMappers
import io.airbyte.commons.json.Jsons
import io.airbyte.protocol.models.v0.AirbyteRecordMessage
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream
import io.github.oshai.kotlinlogging.KotlinLogging
import java.io.IOException
import java.io.PrintWriter
import java.nio.charset.StandardCharsets
import java.sql.Timestamp
import java.util.*

private val LOGGER = KotlinLogging.logger {}

class GcsJsonlWriter(
    config: GcsDestinationConfig,
    s3Client: AmazonS3,
    configuredStream: ConfiguredAirbyteStream,
    uploadTimestamp: Timestamp
) : BaseGcsWriter(config, s3Client, configuredStream), DestinationFileWriter {
    private val uploadManager: StreamTransferManager
    private val outputStream: MultiPartOutputStream
    private val printWriter: PrintWriter
    override val fileLocation: String
    override val outputPath: String

    init {
        val outputFilename: String =
            BaseGcsWriter.Companion.getOutputFilename(uploadTimestamp, FileUploadFormat.JSONL)
        outputPath = java.lang.String.join("/", outputPrefix, outputFilename)

        fileLocation = String.format("gs://%s/%s", config.bucketName, outputPath)
        LOGGER.info {
            "Full GCS path for stream '${stream.name}': ${config.bucketName}/$outputPath"
        }

        this.uploadManager = create(config.bucketName, outputPath, s3Client).get()

        // We only need one output stream as we only have one input stream. This is reasonably
        // performant.
        this.outputStream = uploadManager.multiPartOutputStreams[0]
        this.printWriter = PrintWriter(outputStream, true, StandardCharsets.UTF_8)
    }

    override fun write(id: UUID, recordMessage: AirbyteRecordMessage) {
        val json = MAPPER.createObjectNode()
        json.put(JavaBaseConstants.COLUMN_NAME_AB_ID, id.toString())
        json.put(JavaBaseConstants.COLUMN_NAME_EMITTED_AT, recordMessage.emittedAt)
        json.set<JsonNode>(JavaBaseConstants.COLUMN_NAME_DATA, recordMessage.data)
        printWriter.println(Jsons.serialize(json))
    }

    @Throws(IOException::class)
    override fun write(formattedData: JsonNode) {
        printWriter.println(Jsons.serialize(formattedData))
    }

    override fun closeWhenSucceed() {
        printWriter.close()
        outputStream.close()
        uploadManager.complete()
    }

    override fun closeWhenFail() {
        printWriter.close()
        outputStream.close()
        uploadManager.abort()
    }

    override val fileFormat: FileUploadFormat
        get() = FileUploadFormat.JSONL

    companion object {

        private val MAPPER: ObjectMapper = MoreMappers.initMapper()
    }
}
